ArtOfIllusion/src/artofillusion/script/ScriptRunner.java

```diff
@@ -153,14 +160,14 @@ public static int displayError(String language, Exception ex)
     return line;
   }
 
-  /** Given the name of a file, determine what language it contains based on the extension. */
-
+  /** Given the name of a file, determine what language it contains based on the extension. 
+   @return null if the language is not recognized*/
   public static String getLanguageForFilename(String filename)
   {
     for (String language : LANGUAGES)
       if (filename.endsWith("."+getScriptEngine(language).getFilenameExtension()))
         return language;
-    throw new IllegalArgumentException("Filename \""+filename+"\" does not match any recognized scripting language.");
+    return null;
```

Comment:
Returning null is generally not what we want to do, especially when it should be fairly rare.

(A lot of AOI code does return null/uses null to signal things. This is old code, and would usually be written differently today. We just haven't gotten around to fixing it yet.)

Not sure if there's a better way than throwing an exception - maybe `Optional()`? I'd have to look closer at where it's actually used.

Reply:
I generally agree about nulls not being what we want to return, but on this particular case, I disagree : the fact that a file extension is not recognized only means no ScriptEngine class matching this language extension is implemented, not that the filename is an "illegal argument". If this method should really throw an exception, it should be a NotImplementedException. But generally speaking, exceptions are not meant to be used as a programmatic logic. In this function, the code was already checking for all possible values and should have taken the "language not found" as a possibility from the start (think switch/case/default). 
Instead of null, a String (such as, say, "UNKNOWN_LANGUAGE" or even "") could be a valid return value, without needing Optionals.

Reply:
Certainly exception-driven logic is not ideal, either. The problem with returning `null` is that it can mean anything, and therefore means nothing. It's also very easy to forget to handle the `null` case. Much better to return a named constant of some sort, or other explicit flag, just to force calling code to handle the case. This is another place where I could imagine an enum or similar being handy.